### PR TITLE
Bug 1865306 - Decouple HomeActivity and ExternalAppBrowserActivity RFC 

### PR DESCRIPTION
**Pull request by t-p-white**
⏳ **Deadline for blocking feedback** is 12th January.

:loudspeaker: PSA: This is best viewed as MD, can be viewed in GitHub selecting 'view file'.

Example of steps 1-4:
1. [Bug 1870231 - Remove no-ops from ExternalAppBrowserActivity](https://github.com/mozilla-mobile/firefox-android/pull/4851)
2. [Bug 1870232 - Delegate the Activity's getNavDirections() functionality](https://github.com/mozilla-mobile/firefox-android/pull/4854)
3. [Bug 1870237 - Moved the HomeActivity 'open to browser' functionality to OpenToBrowser.kt and removed the explicit dependency to HomeActivity.kt ](https://github.com/mozilla-mobile/firefox-android/pull/4857)
4. [Bug 1870613 - Extracted the HomeActivitygetBreadcrumbMessage, getIntentSource and getIntentSessionId as generic Activity functions](https://github.com/mozilla-mobile/firefox-android/pull/4872)

The goal of this RFC is primarily to document the _coupled_ nature of the Custom Tabs feature implementation, specifically, the coupling of `HomeActivity` & `ExternalAppBrowserActivity` and to present a solution on how to initiate decoupling.

This RFC is not a fully formed solution to a complete decoupling, but serves to: 
- document the details of the coupling
- document the issues of the coupling 
- provide a guide on initial tasks

This work also has potential overlap with: 

1. [Add `lib-state` alternative to `HomeActivity::openToBrowserAndLoad`](https://bugzilla.mozilla.org/show_bug.cgi?id=1845409)
2. [[lib-state] Investigate ThemeManager and refactor if appropriate](https://bugzilla.mozilla.org/show_bug.cgi?id=1867689)

@MatthewTighe should we include these as part of this project Meta?

### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/shared/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Breaking Changes**: If this is a breaking Android Components change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
7. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
8. Click on the `build-apk-{fenix,focus,klar}-debug` task you're interested in.
9. Click on `View task in Taskcluster` in the new `DETAILS` section.
10. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
https://bugzilla.mozilla.org/show_bug.cgi?id=1865306

### CODE REVIEW

**Review by MatthewTighe — COMMENTED** [review]
Apologies if this feedback is coming a little earlier than you wanted it, but I stumbled on this today and thought I'd drop some thoughts/advice while I was here since I've done a few of these at this point

**Comment by t-p-white** [comment]
> Apologies if this feedback is coming a little earlier than you wanted it, but I stumbled on this today and thought I'd drop some thoughts/advice while I was here since I've done a few of these at this point

Thanks @MatthewTighe! So this version is a little out of date as I have been working on a local version as to still easily reference my code changes for this in another commit, however some of the suggestions are still appropriate and I appreciate the early feedback for sure!

File: docs/rfcs/0011-decouple-home-activity-and-external-activity.md [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+## Summary
+
+The Custom Tabs feature is partially implemented by `ExternalAppBrowserActivity`. `ExternalAppBrowserActivity` has a deeply nested dependency architecture and a tight coupling with the `HomeActivity`.
+
+## Motivation
+
+1. Fenix has [numerous open bugs](https://bugzilla.mozilla.org/show_bug.cgi?id=1815918) and features requests relating to Custom Tabs. This may be partly due to the current Custom Tabs implementation which is not user-friendly to work with. The dependency hierarchy of `ExternalAppBrowserActivity` and close coupling to the `HomeActivity` may be causing difficult to detect bugs, unexpected behavior and generally make an IC reluctant to change these areas.
+1. `HomeActivity` itself has become bloated and Monolithic. It contains unused fields & params, incorrectly labeled `open` functions and generally in need of some refactoring. These issues may be partly due to it's own downstream dependencies. See [0010-add-state-based-navigation.md](https://github.com/mozilla-mobile/firefox-android/blob/1c8de2d2fa7226f568cd49b2161bb9b5471cbcb5/docs/rfcs/0010-add-state-based-navigation.md) for reading around the navigation issue which includes `HomeActivity`.
```
[the remark this answers is not in the document]
Comment (t-p-white) [reply docs/rfcs/0011-decouple-home-activity-and-external-activity.md:17]: Is it the use of both of these terms? It's a fair comment, when I wrote it I was wondering if anyone on the team would disagree to this description (my naive assumption was no) but I'm happy to redact if it feels too opinionated?

**Comment by t-p-white** [comment]
I'm starting to think we may actually need a top-level Activity such as `FenixActivity` to maintain state such as `browsingModeManager` & `themeManager`. The Fenix architecture is very much designed around a single Activity application. This would still provide the decoupling from `HomeActivity` but introduce a new coupling. I think this is still at a good way to start decoupling, refactor and generally clean up some code

**Comment by MatthewTighe** [comment]
> The more I've pulled this apart, the more I'm starting to think we may actually need a top-level Activity such as FenixActivity to maintain the various states such as browsingModeManager & themeManager

It is worth noting that I would like to move both of these examples of state management out of the activity and into the Store. I have an [open patch] for browsingModeManager and am planning to follow up with [themeManager](https://bugzilla.mozilla.org/show_bug.cgi?id=1867689).

In general, I am a little curious about which behaviors of the two activities are shared and which are different. Having a clear picture of those two elements seems like it could help inform the plan to decouple them

**Comment by t-p-white** [comment]
> In general, I am a little curious about which behaviors of the two activities are shared and which are different. Having a clear picture of those two elements seems like it could help inform the plan to decouple them. I don't think that all those examples necessarily need to be added to the RFC though.

Yeah for sure, afaict it looks like:

In terms of state, both Activity's use:
- themeManager: ThemeManager
- browsingModeManager: BrowsingModeManager

Shared functions from `HomeActivity`:
- fun getBreadcrumbMessage(destination: NavDestination): String
- fun getIntentSource(intent: SafeIntent): String?
- fun getIntentSessionId(intent: SafeIntent): String?
- fun getNavDirections(activity: FragmentActivity, from: BrowserDirection, customTabSessionId: String? = null): NavDirections?

Both Activity's override the following `Activity` funcs:
- onProvideAssistContent
- onDestroy
- onResume

**Comment by t-p-white** [comment]
Thanks for your thoughts @MatthewTighe 🙏 

Having spent some time now pulling this apart in various ways, I've come to the conclusion that:

My initial assertion of `ExternalAppBrowserActivity is not a HomeActivity` is ultimately wrong - it is. I think this is partly down to:
1. the naming, it's not a 'home' Activity, it is _the_ Activity (hence my suggestion in step 7)
2. the unclear delineation of responsibility in `HomeActivity`
3. general housekeeping

So - reboot, attempt no. 2. I'm re-framing this RFC as a work package to _reduce_ the decoupling and establish clearer responsibilities for both `Activity`s. This still has the same general principle in mind, but we'll work _with_ the single Activity 

> It seems like a good idea to track as part of this project's Meta as well as the [Home Screen Redux refactor Meta](https://bugzilla.mozilla.org/show_bug.cgi?id=1857657). I am not sure if we can link multiple Epics in Jira though.
Let's sync on this

File: docs/rfcs/0011-decouple-home-activity-and-external-activity.md [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+* Explore alternatives to the current Inheritance structure. E.g. Kotlin Delegation methods, or an alternative 'base' `Activity` model.
+* Prevent unnecessary checks being carried out by `ExternalAppBrowserActivity` in `HomeActivity` lifecycle related functions.
+
+## Drawbacks
+* Current reported Custom Tabs Bugzilla bugs may be affected by the changes mentioned here.
+* No immediate tangible user facing improvements.
+
+## Rationale and alternatives
+### Rationale 
+* It is well established best practice to ['Prefer composition over inheritence'](https://en.wikipedia.org/wiki/Composition_over_inheritance).
```
Comment (jonalmeida) [inline docs/rfcs/0011-decouple-home-activity-and-external-activity.md:157]: This is a building design block of AC as well and why we write all our feature classes that way. Here is [an example](https://github.com/mozilla-mobile/reference-browser/pull/2338) where we are moving some reusable code into that pattern as well.

We can then upstream the change with easily if it's useful.

**Review by jonalmeida — APPROVED** [review]
Nice! Looks good with a few nits.

File: docs/rfcs/0011-decouple-home-activity-and-external-activity.md [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+permalink: /rfc/0011-decouple-home-activity-and-external-activity
+---
+
+* Start date: 2023-12-06
+* RFC PR: [#4732](https://github.com/mozilla-mobile/firefox-android/pull/4732)
+
+## Summary
+The Custom Tabs feature is partially implemented by `ExternalAppBrowserActivity` and `AuthCustomTabActivity`. The use of inheritance in these `Activity`s contributes to a closely coupled system.
+
+The ultimate aim of this RFC is to move the application to a single `Activity` architecture as is [recommended for Modern Android Development (MAD)](https://developer.android.com/topic/architecture/recommendations#ui-layer).
```
[the remark this answers is not in the document]
Comment (t-p-white) [reply docs/rfcs/0011-decouple-home-activity-and-external-activity.md:13]: > I don't think should be our "ultimate aim" but it's a direction we can strive towards. 
👍 
> moving things to a single Activity model for the sake of fitting an architectural guideline is not beneficial.
Agreed! Given the `ExternalAppBroswerActivity` though, I do wonder (perhaps naively) what the benefit of a separate `Activity` is, as it looks to be behaving as a `HomeActivity` albeit adding some checks in `onResume` and `onDestroy`.

[the remark this answers is not in the document]
Comment (jonalmeida) [reply docs/rfcs/0011-decouple-home-activity-and-external-activity.md]: I don't recall any more, but looking at Reference Browser, I'm guessing it was a balance of code readability and simplicity - the `HomeActivity` was the equivalent of the `BrowserActivity` in R-B and very simple in it's nature so the only differences were to override one or two methods. Fenix was iterating too quickly at the time, to stop and make sense of the logic.

File: docs/rfcs/0011-decouple-home-activity-and-external-app-browser-activity.md [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+* `shouldShowOnboarding()`
+* perform checks for Homepage Contile feature activity
+* perform checks for Pocket feature activity
+
+This list is not exhaustive but gives an insight into the redundant work being forced upon the `ExternalAppBrowserActivity`. This applies to many of the other `HomeActivity` lifecycle related functions.
+
+## Proposal
+Reduce the coupling of the `Activity`s. Establish `HomeActivity` as the *main* `Activity` and define clear responsibilities. Below is an list of suggested refactorings to initiate the decoupling.
+
+1. Remove the no-ops from `ExternalAppBrowserActivity`. [Example PR.](https://github.com/mozilla-mobile/firefox-android/pull/4851)
```
[the remark this answers is not in the document]
Comment (t-p-white) [reply docs/rfcs/0011-decouple-home-activity-and-external-app-browser-activity.md:132]: Good spot @boek! I've gated these functions in `HomeActivity`, please see the [updated PR](https://github.com/mozilla-mobile/firefox-android/pull/4851)

File: docs/rfcs/0011-decouple-home-activity-and-external-activity.md [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+1. A close coupling of `ExternalAppBrowserActivity` & `AuthCustomTabActivity` with the `HomeActivity`, which increases the likelihood of:
+   * difficult to detect bugs
+   * difficultly debugging
+   * unexpected behavior
+   * making the area resistant to change
+   * confusion of responsibility
+
+2. `HomeActivity` has a mix of responsibilities, is bloated and monolithic. It contains unused fields, some incorrect visibility modifiers and generally in need of some house-keeping. These issues may be partly due to it's own downstream dependencies. See [0010-add-state-based-navigation.md](https://github.com/mozilla-mobile/firefox-android/blob/1c8de2d2fa7226f568cd49b2161bb9b5471cbcb5/docs/rfcs/0010-add-state-based-navigation.md) for reading around the navigation issue which includes `HomeActivity`.
+
+3. Inheritance has potential to introduce confusion, for example `ExternalAppBrowserFragment` already has duplicate dependencies defined for `UserInteractionHandler`.
```
[the remark this answers is not in the document]
Comment (t-p-white) [reply docs/rfcs/0011-decouple-home-activity-and-external-activity.md]: As discussed, `UserInteractionHandler` is already depending on `UserInteractionHandler` via its dependency on `BaseBrowserFragment`.

However, this conversation also highlighted that `ExternalAppBrowserFragment` does not override `onBackPressed` as the Reference Browser does.

**Comment by t-p-white** [comment]
Latest push was minor formatting changes

**Review by MatthewTighe — COMMENTED** [review]
There are a couple references in your example PRs to consolidating the Activities. For example, [here](https://github.com/mozilla-mobile/firefox-android/pull/4851#discussion_r1444898815) and [here](https://github.com/mozilla-mobile/firefox-android/pull/4872/files#r1445003055). Does this indicate a change to your proposal at all, or is the RFC still ready for final comments?

**Comment by t-p-white** [comment]
> There are a couple references in your example PRs to consolidating the Activities. For example, [here](https://github.com/mozilla-mobile/firefox-android/pull/4851#discussion_r1444898815) and [here](https://github.com/mozilla-mobile/firefox-android/pull/4872/files#r1445003055). Does this indicate a change to your proposal at all, or is the RFC still ready for final comments?

> The next steps are less cleary defined and require more investigation on the completion of the above tasks.

@MatthewTighe good spot! So I would class those comments under this part of the RFC as it's definitely an option I want to explore, but it's by no means tho only option. I realize my wording in those comments was perhaps a little too opinionated on this particular approach, probably due to prototyping at the time

**Review by MatthewTighe — APPROVED** [review]
Plan as stated here seems good to me! Appreciate all the examples, and am curious to see where we go after this. 

I'm glad that the process for this was detailed, as the exploration and iteration seems to have really informed the output. That said, I'm also not sure if the changes planned necessarily would've required an RFC had we known what they were going into it, compared to a final plan on how to reconcile the Activity hierarchy.  Hindsight is 20/20 of course, but it might make for an interesting bit of discussion at some point

**Pull request merged**